### ISSUE TITLE
has function behaves inconsistently

### ISSUE THREAD

**Issue by nik9000** [thinkaurelius/titan#906]
I'm seeing inconsistent behavior on the `has` function:

``` groovy
gremlin> g.V.has('P106link', 'Q36180').map('P106link')[0]
==>{P106link=[Q36180, Q6625963]}
gremlin> g.V.has('P106link', 'Q36180').has('P106link', 'Q36180').map('P106link')[0]
==>{P106link=[Q36180, Q6625963]}
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q36180').map('P106link')[0]
<throws exception because there are no results>
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q6625963').map('P106link')[0]
<throws exception because there are no results>
```

I _think_ what is going on here is that the initial `has` evaluation uses an index which uses 'in' semantics for equals.  I imagine the second line works because of some duplicate condition logic.  I believe the third line fails because it can't use an index and so instead `has` is exact equals semantics.

This is going to be a pain to work around especially because there doesn't seem to be a T.contains operator for `has` so you have to use a closure to check this condition.  That's a pain because, at least conceptually, you could farm all of these queries to a search backend like Elasticsearch which could efficiently perform the required AND operations.

**Comment by smalyshev** [comment]
Some more experiments on this:

```
gremlin> g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q36180').map('P106link')[0]
==>{P106link=[Q214917, Q36180, Q482980, Q49757]}
gremlin> g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q36180').map('P106link')[0].toString()
==>[GremlinStartPipe, GraphQueryPipe(has,vertex), IdentityPipe, IdentityPipe, IdentityPipe, PropertyMapPipe, RangeFilterPipe(0,0)]
```

But:

```
gremlin> g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').filter{true}.has('P106link', 'Q36180').map('P106link')[0].toString()
==>[GremlinStartPipe, GraphQueryPipe(has,vertex), IdentityPipe, IdentityPipe, FilterFunctionPipe, PropertyFilterPipe(P106link,EQUAL,Q36180), PropertyMapPipe, RangeFilterPipe(0,0)]
gremlin> g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').filter{true}.has('P106link', 'Q36180').map('P106link')[0]
(...does not return anything, just hangs....)
```

So looks like when using the optimizer, has() uses IN semantics on SET fields, but where it does PropertyFilterPipe, it uses EQUAL semantics.

**Comment by smalyshev** [comment]
Just to be clear - P???link are SET fields :)

**Comment by nik9000** [comment]
My examples also only use the builtin indexing - @smalyshev has some indexing pushed to Elasticsearch.

**Comment by nik9000** [comment]
This is Titan 0.5.2 as well.

**Comment by dalaro** [comment]
I tried the following script on 0.5.2 and titan05 (0.5.3-SNAPSHOT), but all the queries return the vertex, so I assume I've misunderstood the schema or the data.

```
g = TitanFactory.open("conf/titan-berkeleydb.properties")
m = g.getManagementSystem()
p106link = m.makePropertyKey('P106link').cardinality(Cardinality.SET).dataType(String.class).make()
p21link  = m.makePropertyKey('P21link').cardinality(Cardinality.SET).dataType(String.class).make()
m.buildIndex('P106index', Vertex.class).addKey(p106link).unique().buildCompositeIndex()
m.buildIndex('P21index', Vertex.class).addKey(p21link).unique().buildCompositeIndex()
m.commit()
g.commit()

v = g.addVertex(null)
v.addProperty('P106link', 'Q36180')
v.addProperty('P106link', 'Q6625963')
v.addProperty('P21link', 'Q6581097')
g.commit()

g.V.has('P106link', 'Q36180').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P106link', 'Q36180').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q36180').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q6625963').map('P106link')[0]
```

Can you please modify this script or provide a different script to reflect the problem in your environment?  I'm still trying to get a handle on the issue.

**Comment by nik9000** [comment]
Sure!

**Comment by nik9000** [comment]
This sequence causes trouble:

``` groovy
g = TitanFactory.open("conf/titan-berkeleydb.properties")
m = g.getManagementSystem()
p106link = m.makePropertyKey('P106link').cardinality(Cardinality.SET).dataType(String.class).make()
p21link  = m.makePropertyKey('P21link').cardinality(Cardinality.SET).dataType(String.class).make()
m.buildIndex('P106index', Vertex.class).addKey(p106link).unique().buildCompositeIndex()
m.buildIndex('P21index', Vertex.class).addKey(p21link).unique().buildCompositeIndex()
m.commit()
g.commit()

v = g.addVertex(null)
v.addProperty('P106link', 'Q36180')
v.addProperty('P106link', 'Q6625963')
v.addProperty('P21link', 'Q6581097')
g.commit()

g.V.has('P106link', 'Q36180').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P106link', 'Q36180').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q36180').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q6625963').map('P106link')[0]
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').filter{true}.has('P106link', 'Q6625963').map('P106link')[0]
```

but it doesn't match up with my original problem.  I'll keep digging to see if I can find that problem.

**Comment by nik9000** [comment]
It looks like

``` gremlin
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q6625963').map('P106link')[0]
```

isn't actually throwing an exception because there are no results.  Its throwing an exception because it times out waiting for cassandra.

It looks like that query turns into this sequence of queries to cassandra:

```
KeySliceQuery(key:  59-108-137-160- 81- 51- 54- 49- 56-176, start:   0, end: 255, limit:8000)
KeySliceQuery(key:   1- 52- 12-137-160- 81- 54- 53- 56- 49- 48- 57-183, start:   0, end: 255, limit:715541)
```

And whenever I tried to debug it by injecting stuff between the has clauses I was getting the inconsistent behavior.

That first query is fast - 146 millis.  The second one looks like it times out.

**Comment by smalyshev** [comment]
It looks like the timeout is the result of the fact that `has('P106link', 'Q6625963')` never matches in the last expression, since it uses "equals" semantics instead of "in" semantics and P106link is a SET field with multiple values, so it can never be equal to a single value. I'm still not sure why this would cause a timeout as opposed to just a very long query eventually finishing, but the difference seems to be in the way "has" is treated.  @dalaro please note in your reproduction filter(true) is not present, so the wrong semantics is never used. I think filter() makes the query be treated as has/PropertyFilterPipe predicate instead of aggregating it into single GraphQueryPipe. Looks like PropertyFilterPipe and GraphQueryPipe treat SET fields differently or the former doesn't know how to deal properly with SET fields.

**Comment by dalaro** [comment]
@smalyshev I think I see what you mean.  Is the following script representative of the problem you observed involving `filter{true}`?

``` groovy
g = TitanFactory.open("conf/titan-berkeleydb.properties")
m = g.getManagementSystem()
p106link = m.makePropertyKey('P106link').cardinality(Cardinality.SET).dataType(String.class).make()
p21link  = m.makePropertyKey('P21link').cardinality(Cardinality.SET).dataType(String.class).make()
m.buildIndex('P106index', Vertex.class).addKey(p106link).unique().buildCompositeIndex()
m.buildIndex('P21index', Vertex.class).addKey(p21link).unique().buildCompositeIndex()
m.commit()
g.commit()

v = g.addVertex(null)
v.addProperty('P106link', 'Q36180')
v.addProperty('P106link', 'Q6625963')
v.addProperty('P21link', 'Q6581097')
g.commit()

"Without filter:" // This returns the sole vertex, as expected
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').has('P106link', 'Q36180').map('P106link')[0]

"With filter:" // This does not return anything, unexpected
g.V.has('P106link', 'Q36180').has('P21link', 'Q6581097').filter{true}.has('P106link', 'Q36180').map('P106link')[0]

```

The script prints this:

```
gremlin> load 906bis.groovy
==>titangraph[berkeleyje:/home/dalaro/tinkerelius/titan/conf/../db/berkeley]
==>com.thinkaurelius.titan.graphdb.database.management.ManagementSystem@2e8bd5a8
==>P106link
==>P21link
==>com.thinkaurelius.titan.graphdb.database.management.TitanGraphIndexWrapper@2532718a
==>com.thinkaurelius.titan.graphdb.database.management.TitanGraphIndexWrapper@47e9b9ae
==>null
==>null
==>v[256]
==>e[1a8-74-1l1][256-P106link->Q36180]
==>e[1b4-74-1l1][256-P106link->Q6625963]
==>e[1c0-74-35x][256-P21link->Q6581097]
==>null
==>Without filter:
==>{P106link=[Q36180, Q6625963]}
==>With filter:
gremlin>
```

The pipeline with `filter{true}` includes a likely-problematic `PropertyFilterPipe(P106link,EQUAL,Q36180)`, like you mentioned.  This pipe does not appear in the pipeline produced by the preceding version of the query that lacks `filter{true}`.

I am not yet certain how or whether the Cassandra timeouts relate to this problem, though my preliminary guess is that the `filter{true}` problem is independent of the timeouts, since it seems to be reproducible on BDB.

**Comment by smalyshev** [comment]
@dalaro yes, this seems to be correct representation. I think it should not be Cassandra-specific, since the main problem seems to be handling of SET properties, not timeouts.

**Comment by mbroecheler** [comment]
The `filter{true}` aborts the query rewriting because Gremlin cannot inspect the filter condition. Hence, the first part of the query (up to filter) is rewritten as an index lookup but the has condition following the filter step is executed as a PropertyFilterPipe which cannot handle SET properties (it does getProperty(key) which returns a set and checks for equality which is false).

This script highlights two issues with the TP2 interface:
1) The query optimizer cannot be extended by the database vendor to inject custom query rewrite logic
2) TP2 does not support SET properties

Both of these have been addressed in TP3 and consequently Titan 0.9. It's still a little ways until that is ready for prime time.
